More clearly recommend Gateway API (#52937)
kubernetes/website: content/en/docs/concepts/services-networking/ingress.md

{{< feature-state for_k8s_version="v1.19" state="stable" >}}
{{< glossary_definition term_id="ingress" length="all" >}}

> {{< warning >}}

Reviewer: Same applies as my previous response here.

> {{< warning >}}

Author: I've reverted both to a note, will resolve this and leave follow up conversation in the larger thread above.

The Ingress API has been frozen.
{{< /warning >}}

<!-- body -->

> 

Author: I covered the idea of including Service in the [thread above](https://github.com/kubernetes/website/pull/52937#discussion_r2464356863). As far as linking to a separate page instead of including inline, I think that would result in more users missing the overall point that they should strongly consider using Gateway API over Ingress. That is inherently unclear from the current docs, and I think it's important for us as a community to provide better and clearer guidance for our users.

> ## Consider using Gateway API

Author: Good call - replaced with "stable"

[Gateway API](/docs/concepts/services-networking/gateway/) is GA and includes
many more features. The following table highlights some of the differences
between the APIs:

> | | Ingress | Gateway API |

Reviewer 2: > What about Service?

Service is a core API, and nodePorts and externalIPs are supported out-of-the-box, but neither is a _good_ API for cluster ingress (and externalIPs are almost certainly about to be deprecated).

`type: LoadBalancer` Services are like simplified Ingress/Gateway but require a cloud provider or a non-cloud-based LoadBalancer implementation (eg, MetalLB).

> | | Ingress | Gateway API |

Reviewer: Gateway API is also part of Kubernetes, just not included when you install.

| - | - | - |
| GA | ✅ Yes | ✅ Yes |
| Basic HTTP path matching | ✅ Yes | ✅ Yes |
| Traffic Splitting | ❌ No | ✅ Yes |
| More Protocols (gRPC, TLS) | ❌ No | ✅ Yes |
| More Matchers (Headers, Query Params, Methods) | ❌ No | ✅ Yes |
| Redirects | ❌ No | ✅ Yes |
| Rewrites | ❌ No | ✅ Yes |

> | TLS Validation | ❌ No | ✅ Yes |

Reviewer 3: Rather than a table, consider a description list. With the table you have to be terse. This might imply that an Ingress can't check a client certificate against a trust anchor (untrue).

There are shortcomings but we need to clearly explain what they are.

Reviewer: It is, however, correct to say that the Ingress API has no way to configure that behavior - that requires annotations.

Author: > It is, however, correct to say that the Ingress API has no way to configure that behavior

💯  that's the main point I'm trying to get across here. I've changed the name of the heading from "Ingress" to "Ingress API" and added the following disclaimer above this table:

```
Note that many implementations extend these APIs with implementation-specific features, this table focuses on portable features that are included directly within each Kubernetes API:
```

Does that make sense?

@danwinship had a similar comment below, merging that thread into this one.

Reviewer 4: Looking over the Netlify preview again after reading what @danwinship and @rikatz, yeah, actually, maybe this whole table should just be replaced with "go check out Gateway API, it can probably do what you were thinking about Ingress for." 🙂

Author: Removed the table, PTAL.

| Cross-Namespace References | ❌ No | ✅ Yes |
| AI Extensions | ❌ No | ✅ [Yes](https://gateway-api-inference-extension.sigs.k8s.io/) |
| Service Mesh Support | ❌ No | ✅ Yes |

> | More Features in Development | ❌ No | ✅ Yes |

Reviewer 3: We avoid statements about the future. Explain instead that Ingress is frozen and that the core API isn't expected to change.

Author: Removed the table, PTAL.

{{< glossary_definition term_id="ingress" length="all" >}}

{{< note >}}
The Ingress API has been frozen.

This means that:

> * The Ingress API will not be removed.

Reviewer 3: ```suggestion
* The Ingress API is generally available, and is subject to the [stability guarantees](/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api) for generally available APIs.
  The Kubernetes project has no plans to remove Ingress from Kubernetes.
```

* The Ingress API is no longer being developed, and will have no further changes
  or updates made to it.
{{< /note >}}

<!-- body -->

## Preferred Option: Gateway API

[Gateway API](/docs/concepts/services-networking/gateway/) is stable and
includes many more features (and implementations). Instead of relying on an
assortment of implementation-specific annotations, Gateway API allows you to
use a broad set of Core APIs with a wide variety of implementations. Unlike
Ingress, Gateway API is still under active development and continues to gain

> new capabilities with each release.

Reviewer 3: This isn't quite right for a page that is a list of ingress controllers.
A note callout earlier in the page would be better here.
```suggestion
```

{{< glossary_definition term_id="ingress" length="all" >}}

> {{< note >}}

Reviewer 4: Here, again, I'd strongly prefer this become a warning.

[deploying applications](/docs/tasks/run-application/run-stateless-application-deployment/),
[configuring containers](/docs/tasks/configure-pod-container/configure-pod-configmap/),

> [managing resources](/docs/concepts/workloads/management/).

Reviewer 3: I disagree with removing this line. It is useful to keep.

Reviewer 5: @lmktfy you were always a strong advocate of official labels and annotations, can you advice on how to keep the content without referencing on specific project annotation (that is not official) ?

If the `ingressClassName` is omitted, a [default Ingress class](#default-ingress-class)
should be defined.

> 

Reviewer 3: Why this removal?

{{% /tab %}}
{{% tab name="Namespaced" %}}

> {{< feature-state for_k8s_version="v1.23" state="deprecated" >}}

Author: 🤦 thought I'd reverted all of those, missed one, fixed now.

{{< /caution >}}

There are some ingress controllers, that work without the definition of a

> default IngressClass. It is recommended though, to specify the default

Reviewer 4: 🤔 Should we just drop the first sentence here?

> When using Ingress, start by setting the default IngressClass:

(If we keep it, the leftover comma after "controllers" needs to go. 🙂)

{{% code_sample file="service/networking/tls-example-ingress.yaml" %}}

{{< note >}}
There is a gap between TLS features supported by various Ingress controllers.

> Please refer to documentation for the Ingress controller you've chosen to

Reviewer 3: ```suggestion
You should refer to the documentation for the ingress controller(s) you've chosen to
```

It's IngressClass and Ingress, but it's "ingress controller" not IngressController.
